File: quantum_foundations/core/observables.py
"""Hermitian observables, eigensystems and expectation values.

Specification reference: §5.6.

Observables are kept deliberately separate from gates (§21): a gate is unitary
and evolves a state, an observable is Hermitian and defines what a measurement
device can read out. The two are different objects even when they share a
matrix, as the Pauli operators do.
"""

import numpy as np

from .gates import X, Y, Z
from .states import validate_state
from .types import (
    HERMITIAN_ATOL,
    ComplexMatrix,
    ComplexVector,
    MeasurementOutcome,
    RealVector,
)
from .validation import QuantumValidationError, validate_hermitian, validate_square_matrix

#: Eigenvalues closer together than this are treated as one degenerate subspace.
DEGENERACY_ATOL = 1e-9


def is_hermitian(matrix: ComplexMatrix, *, atol: float = HERMITIAN_ATOL) -> bool:
    """True when A† = A within ``atol``."""
    try:
        validate_hermitian(matrix, atol=atol)
    except QuantumValidationError:
        return False
    return True


def eigensystem(observable: ComplexMatrix) -> tuple[RealVector, ComplexMatrix]:
    """Return ``(eigenvalues, eigenvectors)`` in ascending eigenvalue order.

    Eigenvectors are the *columns* of the returned matrix. Because the
    observable is Hermitian the eigenvalues are real, and are returned as a
    real array rather than as complex numbers with vanishing imaginary parts.
    """
    validate_hermitian(observable)
    values, vectors = np.linalg.eigh(np.asarray(observable, dtype=np.complex128))
    return np.asarray(values, dtype=np.float64), np.asarray(vectors, dtype=np.complex128)


def expectation_value(state: ComplexVector, observable: ComplexMatrix) -> float:
    """⟨A⟩ = ⟨ψ|A|ψ⟩, which is real for a Hermitian observable."""
    array = np.asarray(state, dtype=np.complex128)
    matrix = np.asarray(observable, dtype=np.complex128)
    validate_state(array)
    validate_hermitian(matrix)
    _require_matching_dimensions(array, matrix)
    return float(np.real(np.vdot(array, matrix @ array)))


def variance(state: ComplexVector, observable: ComplexMatrix) -> float:
    """(ΔA)² = ⟨A²⟩ − ⟨A⟩².

    Clamped at zero so that floating-point noise on a sharp eigenstate cannot
    produce a small negative variance.
    """
    matrix = np.asarray(observable, dtype=np.complex128)
    mean = expectation_value(state, matrix)
    mean_of_square = expectation_value(state, matrix @ matrix)
    return float(max(mean_of_square - mean**2, 0.0))


def projective_measurement_distribution(
    state: ComplexVector,
    observable: ComplexMatrix,
) -> list[MeasurementOutcome]:
    """Outcomes of measuring ``observable``, one entry per distinct eigenvalue.

    Degenerate eigenvalues are grouped into a single outcome with the projector
    onto the whole eigenspace, which is what a real measurement device
    distinguishes. Outcomes are ordered by descending eigenvalue so that the
    familiar +1 result is listed first.
    """
    array = np.asarray(state, dtype=np.complex128)
    matrix = np.asarray(observable, dtype=np.complex128)
    validate_state(array)
    validate_hermitian(matrix)
    _require_matching_dimensions(array, matrix)

    values, vectors = eigensystem(matrix)
    outcomes: list[MeasurementOutcome] = []
    start = 0
    for index in range(1, len(values) + 1):
        if index < len(values) and abs(values[index] - values[start]) <= DEGENERACY_ATOL:
            continue
        subspace = vectors[:, start:index]
        projected = subspace @ (subspace.conj().T @ array)
        probability = float(np.clip(np.real(np.vdot(projected, projected)), 0.0, 1.0))
        collapsed = (
            projected / np.sqrt(probability) if probability > 0.0 else np.zeros_like(array)
        )
        outcomes.append(
            MeasurementOutcome(
                eigenvalue=float(np.mean(values[start:index])),
                probability=probability,
                state=collapsed,
            )
        )
        start = index
    return sorted(outcomes, key=lambda outcome: -outcome.eigenvalue)


def spin_observable(direction: RealVector | tuple[float, float, float]) -> ComplexMatrix:
    """σ_n = n_x X + n_y Y + n_z Z for a unit vector ``n`` in ℝ³.

    This is the observable a Stern–Gerlach device measures when its axis points
    along ``n`` (§8.2), and its eigenvalues are ±1 for any unit ``n``.
    """
    vector = np.asarray(direction, dtype=np.float64)
    if vector.shape != (3,):
        raise QuantumValidationError(
            "INVALID_SPIN_AXIS",
            f"A measurement axis must be a 3-vector; received shape {vector.shape}.",
            shape=list(vector.shape),
        )
    if not np.all(np.isfinite(vector)):
        raise QuantumValidationError(
            "NON_FINITE_VALUE", "The measurement axis contains a non-finite value."
        )
    norm = float(np.linalg.norm(vector))
    if abs(norm - 1.0) > 1e-9:
        raise QuantumValidationError(
            "MEASUREMENT_AXIS_NOT_UNIT",
            "A measurement axis must be a unit vector.",
            norm=norm,
        )
    return np.asarray(vector[0] * X + vector[1] * Y + vector[2] * Z, dtype=np.complex128)


def axis_from_angles(theta: float, phi: float) -> RealVector:
    """Cartesian unit vector for the spherical angles used by the UI (§6)."""
    return np.array(
        [
            np.sin(theta) * np.cos(phi),
            np.sin(theta) * np.sin(phi),
            np.cos(theta),
        ],
        dtype=np.float64,
    )


def _require_matching_dimensions(state: ComplexVector, matrix: ComplexMatrix) -> None:
    validate_square_matrix(matrix, label="observable")
    if matrix.shape[0] != state.shape[0]:
        raise QuantumValidationError(
            "OPERATOR_DIMENSION_MISMATCH",
            f"An operator of dimension {matrix.shape[0]} cannot act on a state of "
            f"dimension {state.shape[0]}.",
            operator_dimension=int(matrix.shape[0]),
            state_dimension=int(state.shape[0]),
        )

File: apps/web/frontend/src/math/observables.ts
/**
 * Hermitian observables for one qubit.
 *
 * Mirrors `quantum_foundations/core/observables.py` (§5.6). Observables are
 * kept separate from gates: a gate is unitary and evolves a state, an
 * observable is Hermitian and says what a device can read.
 *
 * Every 2×2 Hermitian matrix can be written
 *
 *     A = c₀I + r·σ,   c₀ = tr(A)/2,   r ∈ ℝ³
 *
 * so its eigenvalues are c₀ ± |r| and its eigenvectors are those of σ_n̂ with
 * n̂ = r/|r|. That gives an exact spectrum in closed form — no numerical
 * diagonalisation, and no eigenvector that wobbles as a slider moves.
 */

import { type Complex, magnitudeSquared } from "./complex";
import { type Matrix, X, Y, Z, addMatrices, identityMatrix, scaleMatrix } from "./gates";
import { type MeasurementAxis, axisEigenstates } from "./measurement";
import { type StateVector, innerProduct } from "./states";

export type HermitianDecomposition = {
  /** The identity component, c₀ = tr(A)/2. */
  offset: number;
  /** The Pauli components (r_x, r_y, r_z). */
  vector: { x: number; y: number; z: number };
  /** |r| — zero exactly when A is a multiple of the identity. */
  length: number;
};

export const decomposeHermitian = (matrix: Matrix): HermitianDecomposition => {
  const a = matrix[0]![0]!;
  const b = matrix[0]![1]!;
  const c = matrix[1]![0]!;
  const d = matrix[1]![1]!;
  const offset = (a.re + d.re) / 2;
  const vector = {
    x: (b.re + c.re) / 2,
    y: (b.im - c.im) / -2,
    z: (a.re - d.re) / 2,
  };
  return { offset, vector, length: Math.hypot(vector.x, vector.y, vector.z) };
};

export type ObservableOutcome = {
  eigenvalue: number;
  probability: number;
  eigenstate: StateVector;
  /** The Bloch direction the eigenstate points along, for drawing. */
  axis: MeasurementAxis;
};

const axisFromVector = (vector: { x: number; y: number; z: number }): MeasurementAxis => {
  const length = Math.hypot(vector.x, vector.y, vector.z) || 1;
  const theta = Math.acos(Math.min(1, Math.max(-1, vector.z / length)));
  const phi = Math.atan2(vector.y, vector.x);
  return { theta, phi: phi < 0 ? phi + 2 * Math.PI : phi };
};

/**
 * The possible readings of `observable` on `state`, with their probabilities
 * and the states that give each reading with certainty.
 *
 * A degenerate observable (a multiple of the identity) has one outcome, which
 * every state returns with certainty.
 */
export const observableOutcomes = (
  observable: Matrix,
  state: StateVector,
): ObservableOutcome[] => {
  const { offset, vector, length } = decomposeHermitian(observable);
  if (length < 1e-12) {
    return [
      {
        eigenvalue: offset,
        probability: 1,
        eigenstate: state,
        axis: { theta: 0, phi: 0 },
      },
    ];
  }
  const axis = axisFromVector(vector);
  const { plus, minus } = axisEigenstates(axis);
  const plusProbability = magnitudeSquared(innerProduct(plus, state));
  const minusProbability = magnitudeSquared(innerProduct(minus, state));
  const total = plusProbability + minusProbability;
  return [
    {
      eigenvalue: offset + length,
      probability: plusProbability / total,
      eigenstate: plus,
      axis,
    },
    {
      eigenvalue: offset - length,
      probability: minusProbability / total,
      eigenstate: minus,
      axis: axisFromVector({ x: -vector.x, y: -vector.y, z: -vector.z }),
    },
  ];
};

/** ⟨A⟩ = ⟨ψ|A|ψ⟩ = Σ pᵢ λᵢ. */
export const expectationValue = (observable: Matrix, state: StateVector): number =>
  observableOutcomes(observable, state).reduce(
    (total, outcome) => total + outcome.eigenvalue * outcome.probability,
    0,
  );

/** (ΔA)² = ⟨A²⟩ − ⟨A⟩², clamped at zero against floating-point noise. */
export const variance = (observable: Matrix, state: StateVector): number => {
  const outcomes = observableOutcomes(observable, state);
  const mean = outcomes.reduce((total, o) => total + o.eigenvalue * o.probability, 0);
  const meanOfSquare = outcomes.reduce(
    (total, o) => total + o.eigenvalue ** 2 * o.probability,
    0,
  );
  return Math.max(0, meanOfSquare - mean ** 2);
};

/** σ_n = n_x X + n_y Y + n_z Z for a unit vector n. */
export const spinObservable = (direction: { x: number; y: number; z: number }): Matrix =>
  addMatrices(
    addMatrices(scaleMatrix(X, direction.x), scaleMatrix(Y, direction.y)),
    scaleMatrix(Z, direction.z),
  );

export const isHermitianMatrix = (matrix: Matrix, atol = 1e-10): boolean => {
  const conjugateEqual = (left: Complex, right: Complex) =>
    Math.abs(left.re - right.re) <= atol && Math.abs(left.im + right.im) <= atol;
  return (
    Math.abs(matrix[0]![0]!.im) <= atol &&
    Math.abs(matrix[1]![1]!.im) <= atol &&
    conjugateEqual(matrix[0]![1]!, matrix[1]![0]!)
  );
};

export const IDENTITY_2 = identityMatrix(2);
